partiql-eval/src/main/kotlin/org/partiql/eval/internal/operator/rel/RelExclude.kt

```diff
@@ -0,0 +1,178 @@
+            // tuple wildcard at current level. return empty struct
+            return structValue<PartiQLValue>()
+        }
+        val attrsToRemove = leavesSteps.filterIsInstance<ExcludeStep.StructField>()
+            .map { it.attr }
+            .toSet()
+        val entriesWithRemoved = structValue.entries.filter { structField ->
+            !attrsToRemove.contains(structField.first)
+        }
+        val finalStruct = entriesWithRemoved.map { structField ->
+            val name = structField.first
+            var expr = structField.second
+            // apply case-sensitive tuple attr exclusions
+            val structFieldCaseSensitiveKey = ExcludeStep.StructField(name, ExcludeFieldCase.SENSITIVE)
+            branches.find {
+                it.step == structFieldCaseSensitiveKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            // apply case-insensitive tuple attr exclusions
+            val structFieldCaseInsensitiveKey = ExcludeStep.StructField(name, ExcludeFieldCase.INSENSITIVE)
+            branches.find {
+                it.step == structFieldCaseInsensitiveKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            // apply tuple wildcard exclusions
+            val tupleWildcardKey = ExcludeStep.StructWildcard
+            branches.find {
+                it.step == tupleWildcardKey
+            }?.let {
+                expr = excludeOnPartiQLValue(expr, it)
+            }
+            Pair(name, expr)
+        }
+        return structValue(finalStruct)
+    }
+
+    /**
+     * Returns a [PartiQLValue] created from an iterable of [coll]. Requires [type] to be a collection type
+     * (i.e. [PartiQLValueType.LIST], [PartiQLValueType.BAG], or [PartiQLValueType.SEXP]).
+     */
+    @OptIn(PartiQLValueExperimental::class)
+    private fun newCollValue(type: PartiQLValueType, coll: Iterable<PartiQLValue>): PartiQLValue {
+        return when (type) {
+            PartiQLValueType.LIST -> listValue(coll)
+            PartiQLValueType.BAG -> bagValue(coll)
+            PartiQLValueType.SEXP -> sexpValue(coll)
+            else -> error("Collection type required")
+        }
+    }
+
+    @OptIn(PartiQLValueExperimental::class)
+    private fun excludeOnCollValue(
+        coll: CollectionValue<*>,
+        type: PartiQLValueType,
+        exclusions: ExcludeNode
+    ): PartiQLValue {
+        val leavesSteps = exclusions.leaves.map { leaf -> leaf.step }
+        val branches = exclusions.branches
+        if (leavesSteps.any { it is ExcludeStep.CollWildcard }) {
+            // collection wildcard at current level. return empty collection
+            return newCollValue(type, emptyList())
+        } else {
+            val indexesToRemove = leavesSteps.filterIsInstance<ExcludeStep.CollIndex>()
+                .map { it.index }
+                .toSet()
```

Comment:
Then this else branch would move to the top level and you wouldn't have to traverse yet again